flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java

```diff
@@ -158,6 +160,32 @@ public AbstractStreamOperatorV2(StreamOperatorParameters<OUT> parameters, int nu
                         environment.getExternalResourceInfoProvider());
     }
 
+    private WatermarkOutputMultiplexer setupWatermarkMultiplexer(
+            StreamStatusMaintainer streamStatusMaintainer, int inputCount) {
+        WatermarkOutputMultiplexer multiplexer =
+                new WatermarkOutputMultiplexer(
+                        new WatermarkOutput() {
+                            @Override
+                            public void emitWatermark(
+                                    org.apache.flink.api.common.eventtime.Watermark watermark) {
+                                try {
+                                    processWatermark(new Watermark(watermark.getTimestamp()));
+                                } catch (Exception e) {
+                                    throw new WrappingRuntimeException(e);
+                                }
+                            }
+
+                            @Override
+                            public void markIdle() {
+                                streamStatusMaintainer.toggleStreamStatus(StreamStatus.IDLE);
```

Comment:
This is the part of the whole solution that I am unsure the most. With this approach the bookkeeping happens both on the operator and stream task level. As far as I can tell it would work with this method empty.

Theoretically we could also move the whole bookkeeping logic from the StreamTask level to the operator level.

Reply:
What logic on the `StreamTask` level are you referring to?

Reply:
Everything that is in the `StreamStatusMaintainer(s)`